### PR TITLE
Fix behavior for nested DeprecationLogger.whileDisabled invocations

### PR DESCRIPTION
**Pull request by donat**
All deprecation warnings should be suppressed in the code executed with  DeprecationLogger.whileDisabled(). However, the current implementation does not handle nested whileDisabled() calls. So, if a build logic has the following pattern

    DeprecationLogger.whileDisabled {
        DeprecationLogger.whileDisabled {
        }
        DeprecationLogger.deprecate("nag")
            .willBeRemovedInGradle9()
            .undocumented()
            .nagUser()
      }

then the nagging message will appear on the console.

This is because the implementation uses a single, static boolean to keep track of whether or not log messages should be suppressed. That boolean will be set to false when the inner whileDisabled() method returns.

This PR fixes this issue by counting how many levels of nested whileDisabled() are there and only disables the suppression after the outermost block finishes.

### CODE REVIEW

**Review by reinsch82 — APPROVED** [review]
LGTM

**Review by eskatos — APPROVED** [review]
:+1: 

There's a test failing because this change correctly hides a warning somewhere.

**Comment by donat** [comment]
@bot-gradle test RFM

**Comment by eskatos** [comment]
@bot-gradle test RFM

**Comment by eskatos** [comment]
@bot-gradle test RFN

**Comment by eskatos** [comment]
@bot-gradle test RFN

**Comment by eskatos** [comment]
@bot-gradle test APT

**Comment by reinsch82** [comment]
@bot-gradle help

**Comment by reinsch82** [comment]
@bot-gradle test APT

**Comment by donat** [comment]
Thanks, @eskatos for wrapping the generated init method with `whileDisabled`.

**Comment by donat** [comment]
The performance regression failure seems valid though.

**Comment by eskatos** [comment]
Yes, this incurs a <1% performance regression on "cold daemon" scenarios. I tried to mitigate by making the changed code more efficient but failed. I think we should accept the regression given it is tiny on non critical scenarios and this fixes a real problem.

**Comment by donat** [comment]
@bot-gradle test RFM

**Review by reinsch82 — APPROVED** [review]
LGTM

**Review by eskatos — APPROVED** [review]

File: subprojects/model-core/src/main/java/org/gradle/internal/instantiation/generator/AsmBackedClassGenerator.java [hunk 1]
```diff
@@ -669,8 +690,28 @@ private void visitDeclaredAnnotationsOf(Constructor<?> constructor, MethodVisito
             }
         }
 
+        /**
+         * Generates the init method with deprecation logging disabled.
+         *
+         * This is necessary because the initialization work invokes property getters on the decorated instance.
```
Comment (big-guy) [inline subprojects/model-core/src/main/java/org/gradle/internal/instantiation/generator/AsmBackedClassGenerator.java:696]: what's an example of this?

Comment (eskatos) [reply subprojects/model-core/src/main/java/org/gradle/internal/instantiation/generator/AsmBackedClassGenerator.java:696]: An extension with a deprecated property with the getter that emits deprecation warnings. We have some in the codebase, e.g. `KotlinPluginOptions.jvmTarget`.

**Comment by donat** [comment]
@bot-gradle test APT

**Pull request merged**